### PR TITLE
Readd VIP after interface down/up

### PR DESCRIPTION
**Pull request by rwgroenenberg**
When a virtual router is defined with one or more VIPs on other interface than the VRRP interface, bringing down and up (i.e. `ifdown <if> ; ifup <if>` ) such interface does remove the VIP, but not re-add it.
E.g. interface eth0 is used for VRRP, VIPs are configured on eth1 and eth2. Bringing down eth2 removes the VIP, but `ifup eth2` does not result in the VIP to be re-added.

(We see this happen in a setup where Keepalived is used for HA monitoring and VIP assignment, and one interface is temporarily brought down for some reconfig without switching from Master to Backup).

This change handles re-adding a VIP when an interface comes available (only when started with the --readd-vips option) and also invokes the notify script with 'READD'.

### CODE REVIEW

**Comment by acassen** [comment]
Hello,
I would rather name the feature as --persistent-vips and would not introduce any new FSM state since it is only consistent for logging and not to drive VRRP FSM.
If I remind properly, this feature is already part of Beta branch, right Quentin ?

regs,
Alexandre

**Comment by rwgroenenberg** [comment]
Hi Alexandre,
I don't mind changing the name.
For our use we do need to invoke the notify script in order to add any IP routes that use the re-added VIP address. Is there another way to do that without the READD 'state' (not really a state).
Cheers, Robert

**Comment by pqarmitage** [comment]
Alexandre, yes the beta branch does re-add VIPs if they are removed, but in a rather different way. If a VIP is removed, the vrrp instance will revert to the BACKUP state, send a priority 0 advert, schedule a time out as though it had received a priority 0 advert, and a new master will be elected within one advert interval of the address being removed.

I was going to suggest that if we can have a final release of the v1.4.x series to bring that fully up-to-date, then I will merge the beta branch into the master branch (we had an update today from @lkarsten that he has been running the beta branch in production for the last three weeks without a problem).

I'm not convinced that this is quite the right way to add the functionality, I'll give a breakdown on this shortly. In the meantime, it would seem sensible to use the beta branch for this functionality until it is merged into the master branch.

**Comment by pqarmitage** [comment]
The following are my thoughts about the readd patch:
- I think the functionality should be configured via the configuration file, and be configurable per vrrp instance, rather than it being a global command line option.
- The patch doesn't include eVIPs
- The trigger to re-adding VIPs is if an address is added, which doesn't seem quite the right logic
- If a VIP or eVIP is deleted, the vrrp instance is not set to backup state and so will continue to advertise the addresses 
- For a large configuration, everytime an address is added, keepalived will have to spin through all vrrp instances and all VIPs to see if it should add any other addresses. This will be a large overhead.
- Likewise, every time an address is deleted, all VIPs on all vrrp instances will be compared.
- When addresses are deleted, the interface that the address was on is not checked to see if it matches the interface keepalived has configured that address on.

I've had a look at the readd patch, and unfortunately it cannot be merged in it's current state since it has a memory leak in it.

I will add some comments against the patch in case that is helpful to anyone.

**Comment by rwgroenenberg** [comment]
Hi Quentin, Alexandre,

Thanks for your comments so far, I'll address the missing check(s) and memory leak, etc. Suggestions for improvement are appreciated.

Regarding the suggestion to go to Backup state when a interface holding a VIP (so not the primary or a tracked) interface goes down, will not work for us. In our use-case (HA cluster) we use various checker scripts to determine whether the application is healthy and also to health of various interfaces and reachability of gateway. The outcome of that affect the effective priority and thus the master election. I understand that that is not the 'core focus' of Keepalived and probably Pacemaker is more suitable for this. The reason that we currently do use Keepalived is mainly because we also deploy the same product in n+1 model, where Keepalived _is_ used to load balance the traffic over the available nodes.

**Comment by rwgroenenberg** [comment]
Hi Quentin,
I fully understand that changes need to work for all users. That's also why this 'feature' is disabled by default and has to be enabled to work. I've done some rework based on your comments and added the eVIP.
(this change was certainly not my best... It was actually created quite a while ago against 1.2.19 and I'd just ported it to master. Should've done a better job at that :/

I'll also change the enabling to the config file.
Still need to think about the notify part (a compile option might help ;)

Normally Keepalived should be the one managing the VIPs. At first sight it seems a bit odd to also do the reverse (VIP presence controls Keepalived state).
Indeed, for our use we would want to be able to disable the implicit tracking of the VIP-interfaces. Actually more users may suffer from such a change in behavior if it becomes the default (and only) option.

File: keepalived/core/keepalived_netlink.c [hunk 1]
```diff
@@ -584,6 +584,12 @@ netlink_if_address_filter(__attribute__((unused)) struct sockaddr_nl *snl, struc
 			if (!ifp->sin6_addr.s6_addr16[0] && ifa->ifa_scope == RT_SCOPE_LINK)
 				ifp->sin6_addr = *(struct in6_addr *) addr;
 		}
+
+		/* Re-add VIP(s) */
+		if (__test_bit(READD_VIPS_BIT, &debug)) {
+			vrrp_handle_ip_change(IF_NAME(ifp), ifa->ifa_family, addr,
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/core/keepalived_netlink.c]: Fixed

File: keepalived/vrrp/vrrp.c [hunk 2]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
...
+	list    l;
+	element e;
+
+	if (!vrrp_data)
+		return 1;
+
+	l = vrrp_data->vrrp;
+	for (e = LIST_HEAD(l); e; ELEMENT_NEXT(e)) {
+		vrrp_t		*vrrp = ELEMENT_DATA(e);
+		list		vips = alloc_list(NULL, dump_ipaddress);
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed

File: keepalived/vrrp/vrrp.c [hunk 3]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
...
+			 */
+			continue;
+		}
+
+		for (ve = LIST_HEAD(vrrp->vip); ve; ELEMENT_NEXT(ve)) {
+			ip_address_t	*ipaddr = ELEMENT_DATA(ve);
+
+			if ((action == IPADDRESS_ADD) && !ipaddr->set) {
+				if ((family == IP_FAMILY(ipaddr)) &&
+				    (strcmp(IF_NAME(ipaddr->ifp), iface) == 0) ) {
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed

File: keepalived/vrrp/vrrp.c [hunk 4]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
 	}
 }
 
+/* Handle IP address or interface changes to add/delete VIP */
+int
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed

File: keepalived/vrrp/vrrp.c [hunk 5]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
...
+		int		n = 0;
+
+		if (!vrrp->vipset) {
+			/* VIPs not set for this VRRP, no need to evaluate VIP address(es)
+			 * for this interface.
+			 */
+			continue;
+		}
+
+		for (ve = LIST_HEAD(vrrp->vip); ve; ELEMENT_NEXT(ve)) {
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed

File: keepalived/vrrp/vrrp.c [hunk 6]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
...
+				    (strcmp(IF_NAME(ipaddr->ifp), iface) == 0) ) {
+					// Match address/mask?
+					list_add(vips, ipaddr);
+					n++;
+				}
+
+			} else if (action == IPADDRESS_DEL) {
+				/* Clear VIP state on delete */
+				if ((family == IP_FAMILY(ipaddr)) &&
+				    (IP_IS6(ipaddr) ? (memcmp((void*)&ipaddr->u.sin6_addr, address,
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed, both interface and address are now checked

File: keepalived/vrrp/vrrp.c [hunk 7]
```diff
@@ -2938,3 +2938,71 @@ clear_diff_script(void)
...
+				{
+					log_message(LOG_INFO, "Clearing VIP %s\n", ipaddresstos(NULL, ipaddr));
+					ipaddr->set = 0;
+				}
+			}
+		}
+
+		/* Add matching VIPs */
+		if (!LIST_ISEMPTY(vips)) {
+			log_message(LOG_INFO, "(re)adding VIP%s:\n", n?"s":"");
```
[the remark this answers is not in the document]
Comment (rwgroenenberg) [reply keepalived/vrrp/vrrp.c]: Fixed, using LIST_SIZE instead local counter

**Comment by rwgroenenberg** [comment]
To give some background why IMHO a VIP/interface going down should not always lead to going into Fault state:
assume two nodes, Active/Standby, with VIPs on multiple interfaces for SCTP multi-homing. When one of the interfaces with a VIP on it goes down, SCTP communication is still possible using the remaining interface. So this node, although somewhat degraded, may still be a better option than the other node where the application is down. Therefor we want the priorities+weights from the track scripts control which node is Master instead of ruling out a node where one interface (or VIP) is unavailable.

**Comment by pqarmitage** [comment]
@rwgroenenberg I have now implemented my planned changes for handling the deletion of VIPs, eVIPs, virtual routes and virtual rules. In addition I have implemented restoring static ip addresses, routes and rules when they are deleted.

In the light of your comment https://github.com/acassen/keepalived/pull/860#issuecomment-387429714 I have added a `no_track` option to each of static and virtual ip addresses, routes and rules. Unless `no_track` is specified against an address/route/rule, action will be taken to restore it if it deleted. If it is a virtual address/route/rule, the vrrp instance will drop back to backup, and if it then subsequently will re-add any virtual addresses/routes/rules when it next transitions to master; if it cannot re-add the item (this will occur if the interface of an address or route is down), then the vrrp instance will go to fault statue until the interface is restored. If it is a static address/route/rule keepalived will attempt to re-add the deleted item.

For static address/routes/rules there is an additional option - named `track_group` entries. A track group has a list of vrrp instances. If a static address/route(/rule) cannot be added (due to an interface being down), those vrrp instances will be put into fault state until the interface is restored.

In summary, by default any address, route or rule will be re-added if it is deleted, unless the address/route/rule is configured with the `no_track` option. In addition, for any static address/route(/rule) vrrp instances can be configured to go to fault state if the address/route/rule cannot be configured.

Could you please check the above functionality, and update this log with a description of whether this meets your needs, and if not what else you need.

**Comment by pqarmitage** [comment]
Closing due to no comment for 4 weeks, and the functionality requested is now supported by keepalived.
